### ISSUE TITLE
issue during upgrade to latest release 1.2.2

### ISSUE THREAD

**Issue by tpokorra** [theCalcaholic/nextcloud-secrets#21]
I got this error on `occ upgrade` this morning:

```
An unhandled exception has been thrown:
TypeError: Argument 1 passed to OCA\Secrets\Migration\Version1001Date20230122142756::fixSerialization() must be of the type string or null, resource given, called in [...]/nextcloud/apps/secrets/lib/Migration/Version1001Date20230122142756.php on line 92 and defined in [...]/nextcloud/apps/secrets/lib/Migration/Version1001Date20230122142756.php:97
```

**Comment by theCalcaholic** [comment]
Hey, thanks for the report. Unfortunately I can't reproduce this issue, but I have tried to come up with a fix. 

Can you try this (hopefully fixed) version of the app and tell me if it worked for you?
https://cloud.knoeppler.org/index.php/s/wfGMnBTXC66Xbqz

How to install:

1. Navigate to your `nextcloud/apps` folder
2. Download the *.tar.gz file to your server: `wget https://cloud.knoeppler.org/index.php/s/wfGMnBTXC66Xbqz/download/secrets-fix-resource-handling.tar.gz -O ./secrets.tar.gz`
3. Remove the existing "secrets" directory: `rm -r ./secrets`
4. Extract the archive: `tar xf secrets.tar.gz`
5. Fix ownership of the secrets folder: `sudo chown -R www-data: ./secrets` (assuming the user of your php process/web service is `www-data` - otherwise, replace with that user)
6. Remove the archive: `rm secrets.tar.gz`

**Comment by tpokorra** [comment]
Thanks for looking into this!

I had uninstalled the app, so that we could use the Nextcloud again.
I have now downloaded your archive, and when I run:

`php occ app:enable secrets` 

I get this:

```
An unhandled exception has been thrown:
TypeError: Argument 1 passed to OCA\Secrets\Migration\Version1001Date20230122142756::fixSerialization() must be an instance of OCA\Secrets\Migration\mixed, resource given, called in [...]/nextcloud/apps/secrets/lib/Migration/Version1001Date20230122142756.php on line 93 and defined in [...]/nextcloud/apps/secrets/lib/Migration/Version1001Date20230122142756.php:98
```

I don't mind any data loss, I think I might just reset the existing secrets.
I would uninstall the app, drop the tables `oc_secrets_id_seq` and `oc_secrets`, and install the app again.

**Comment by theCalcaholic** [comment]
That will work. If you want to help me fix this, you could give me an export of your (end to end encrypted) secrets, or at least the `iv` column

You can get that with `select iv from oc_secrets`

That would help me immensely to further debug the issue

**Comment by tpokorra** [comment]
I started to wonder if the reason for this issue is that we are using PostgreSQL instead of MySQL as the database backend. I will test this in another test setup.

Here is our `iv` column.

```
select iv from oc_secrets;
                      iv                      
----------------------------------------------
 \xc2bac280c38f49c2b5c299c3bd48c380c2afc28264
 \x3e1dc3bfc2bec28128c3a2c2996fc3986230
 \x4ac2ad0dc29a0cc296432a5d0ac294c29a
```

**Comment by tpokorra** [comment]
@theCalcaholic Yes this is probably all related to the pgsql database.
On a fresh installation, when I create secrets, I later cannot load them via the secret link, I get:

```
Cannot assign resource to property OCA\\Secrets\\Db\\Secret::$encrypted of type ?string in file '[...]/nextcloud/lib/public/AppFramework/Db/Entity.php' line 134
```

It works on a mysql database.

Question is: should I try to work on a pull request to make it work for postgresql?

**Comment by theCalcaholic** [comment]
@tpokorra Thank you, that's very insightful. I believe, I know a workaround for this (don't use the blob type for DB fields, which is making me a lot of trouble anyways). This type was being used for the `iv` and the `encrypted` field, which both seem to fail on your Postgres instance.

I'll see if I can reproduce this (which, with a postgres db will hopefully be easy) and create a fix. Thanks for offering a PR, but I think, I can do it myself today or tomorrow, so that should be fine :)

**Comment by saveip** [comment]
We have got the same problem as @tpokorra using postgresql. Hopefully you can get it fixed to run on postgresql too !

**Comment by theCalcaholic** [comment]
Should now be fixed in version [1.3.0](https://github.com/theCalcaholic/nextcloud-secrets/releases/tag/v1.3.0). Please report, if you encounter any issues with the new release.

**Comment by tpokorra** [comment]
Thank you!
After deleting the rows in `oc_secrets`, I was able to install v1.3.0, and it works fine on Postgresql!
